src/smackw32/smackw32.c

```diff
@@ -9,55 +9,236 @@
+
+    ma_paged_audio_buffer_page *newPage;
+    result_ma = ma_paged_audio_buffer_data_allocate_page(&user->miniaudio.paged_audio_buffer_data, framesOut, NULL, NULL, &newPage);
+    if (result_ma != MA_SUCCESS) {
+        fprintf(stderr, "ma_paged_audio_buffer_data_allocate_page failed");
+        return;
+    }
+
+    ma_data_converter_process_pcm_frames(&user->miniaudio.data_converter,
+        audio_data,
+        &framesIn,
+        newPage->pAudioData,
+        &framesOut);
+    result_ma = ma_paged_audio_buffer_data_append_page(&user->miniaudio.paged_audio_buffer_data, newPage);
+    if (result_ma != MA_SUCCESS) {
+        fprintf(stderr, "ma_paged_audio_buffer_data_append_page failed");
+        ma_paged_audio_buffer_data_free_page(&user->miniaudio.paged_audio_buffer_data, newPage, NULL);
+    }
+    if (!ma_sound_is_playing(&user->miniaudio.sound)) {
+        ma_sound_seek_to_pcm_frame(&user->miniaudio.sound, frame_length_at_start);
+
+        ma_sound_start(&user->miniaudio.sound);
+    }
+}
+
+static void copy_palette(Smack* smack) {
+    SmackUserData *const user = smack->user;
+
+    const unsigned char* pal = smk_get_palette(user->smk_handle);
     memcpy(smack->Palette, pal, 256 * 3);
 }
 
 Smack* SmackOpen(const char* name, uint32_t flags, uint32_t extrabuf) {
     double usf;
-    Smack* smack;
-    double fps;
+    Smack* smack = NULL;
+    SmackUserData *user = NULL;
 
     smk smk_handle = smk_open_file(name, SMK_MODE_MEMORY);
     if (smk_handle == NULL) {
-        return NULL;
+        goto error;
     }
 
     smack = malloc(sizeof(Smack));
+    if (smack == NULL) {
+        goto error;
+    }
+    memset(smack, 0, sizeof(*smack));
+
+    user = malloc(sizeof(SmackUserData));
+    if (user == NULL) {
+        goto error;
+    }
+    memset(user, 0, sizeof(*user));
+    smack->user = user;
 
     // libsmacker doesn't tell us whether the palette is new on each frame or not, so just assume it always is new
     smack->NewPalette = 1;
 
     // smk_handle is added to hold a pointer to the underlying libsmacker instance
-    smack->smk_handle = smk_handle;
+    user->smk_handle = smk_handle;
 
     smk_info_all(smk_handle, NULL, &smack->Frames, &usf);
-    fps = 1000000.0 / usf;
-    smack->MSPerFrame = (1 / fps) * 1000;
+    smack->MSPerFrame = (unsigned long)(usf / 1000);
     smk_info_video(smk_handle, &smack->Width, &smack->Height, NULL);
     smk_enable_video(smk_handle, 1);
 
+    audio_init(smack);
+
     if (smk_first(smk_handle) == SMK_ERROR) {
-        smk_close(smk_handle);
-        free(smack);
-        return NULL;
+        goto error;
     }
     copy_palette(smack);
     return smack;
+error:
+    if (smk_handle) {
+        smk_close(smk_handle);
+    }
+    free(user);
+    free(smack);
+    return NULL;
 }
 
 int SmackSoundUseDirectSound(void* dd) {
-    // TODO: do some miniaudio init
+    // FIXME: is this a more appropriate place to init miniaudio? What is type of argument?
```

Comment:
Because the arguent of `SmackSoundUseDirectSound` was `NULL`, I did the audio initialization in `SmackOpen`.